controllers/templatesync/template_sync.go

```diff
@@ -523,11 +531,89 @@ func (r *PolicyReconciler) Reconcile(ctx context.Context, request reconcile.Requ
 		policySystemErrorsCounter.WithLabelValues(instance.Name, "", "client-error").Inc()
 	}
 
+	err = r.cleanUpExcessTemplates(ctx, dClient, *instance, parentTemplates)
+	if err != nil {
+		resultError = err
+		reqLogger.Error(resultError, "Error cleaning up templates")
+	}
+
 	reqLogger.Info("Completed the reconciliation")
 
 	return reconcile.Result{}, resultError
 }
 
+// cleanUpExcessTemplates compares existing policy templates on the cluster to those contained in the policy,
+// and deletes those that have been renamed or removed from the parent policy
+func (r *PolicyReconciler) cleanUpExcessTemplates(
+	ctx context.Context,
+	dClient dynamic.Interface,
+	instance policiesv1.Policy,
+	parentTemplates []string,
+) error {
+	// build list of policy template resources to check
+	tmplCRDs, err := dClient.Resource(schema.GroupVersionResource{
```

Comment:
Could you please take advantage of caching by using the client associated with the `PolicyReconciler`? Additionally, could you please configure the cache with the label selector similar to what the config-policy-controller does?

https://github.com/open-cluster-management-io/config-policy-controller/blob/8203f5fb9e5fbd72b85852d54f7bd91331c298ce/main.go#L180-L187

Reply:
@mprahl I'm working on this but am a bit unfamiliar with using the cache. I'm confused on choosing between v1 and v1beta1 - it seems like the cache selectors function you linked automatically chooses the right one, but when I use the reconciler client to list the CRDs, the controller needs to pick between v1 and v1beta1 for the object list it is populating with the CRDs, right? Like this:

```
	crds := extensionsv1.CustomResourceDefinitionList{}
	label, err := labels.NewRequirement(
		"policy.open-cluster-management.io/policy-type",
		selection.Equals,
		[]string{"template"},
	)
	err = r.List(ctx, &crds, &client.ListOptions{
		LabelSelector: labels.NewSelector().Add(*label),
	})
```

Unless there is a data type I'm not aware of that accepts a list of either v1 or v1beta1 CRDs, I'm not sure how to use the reconciler client here without a duplicate block that lists v1beta1 CRDs if the above block fails.

Reply:
Here is roughly what I had in mind:
```diff
diff --git a/controllers/templatesync/template_sync.go b/controllers/templatesync/template_sync.go
index 8036214..85d822e 100644
--- a/controllers/templatesync/template_sync.go
+++ b/controllers/templatesync/template_sync.go
@@ -12,11 +12,14 @@ import (
        "github.com/go-logr/logr"
        "github.com/prometheus/client_golang/prometheus"
        depclient "github.com/stolostron/kubernetes-dependency-watches/client"
+       extensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+       extensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
        "k8s.io/apimachinery/pkg/api/equality"
        "k8s.io/apimachinery/pkg/api/errors"
        "k8s.io/apimachinery/pkg/api/meta"
        metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
        "k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+       "k8s.io/apimachinery/pkg/labels"
        "k8s.io/apimachinery/pkg/runtime"
        "k8s.io/apimachinery/pkg/runtime/schema"
        "k8s.io/apimachinery/pkg/types"
@@ -36,8 +39,9 @@ import (
 )
 
 const (
-       ControllerName string = "policy-template-sync"
-       policyFmtStr   string = "policy: %s/%s"
+       ControllerName  string = "policy-template-sync"
+       policyFmtStr    string = "policy: %s/%s"
+       PolicyTypeLabel string = "policy.open-cluster-management.io/policy-type"
 )
 
 var log = ctrl.Log.WithName(ControllerName)
@@ -551,30 +555,36 @@ func (r *PolicyReconciler) cleanUpExcessTemplates(
        parentTemplates []string,
 ) error {
        // build list of policy template resources to check
-       tmplCRDs, err := dClient.Resource(schema.GroupVersionResource{
-               Group:    "apiextensions.k8s.io",
-               Version:  "v1",
-               Resource: "customresourcedefinitions",
-       }).List(ctx, metav1.ListOptions{
-               LabelSelector: "policy.open-cluster-management.io/policy-type=template",
-       })
-       if err != nil {
-               return fmt.Errorf("error getting policy template CRDs: %w", err)
-       }
-
+       crdLabelSelector := labels.SelectorFromSet(map[string]string{PolicyTypeLabel: "template"})
+       crdsv1 := extensionsv1.CustomResourceDefinitionList{}
        tmplGVRs := []schema.GroupVersionResource{}
 
-       for _, crd := range tmplCRDs.Items {
-               versions, ok := crd.Object["spec"].(map[string]interface{})["versions"].([]interface{})
-               if !ok {
-                       return fmt.Errorf("template CRD is formatted incorrectly: no versions found")
+       err := r.List(ctx, &crdsv1, &client.ListOptions{LabelSelector: crdLabelSelector})
+       if err == nil {
+               for _, crd := range crdsv1.Items {
+                       tmplGVRs = append(tmplGVRs, schema.GroupVersionResource{
+                               Group:    crd.Spec.Group,
+                               Resource: crd.Spec.Names.Plural,
+                               // TODO: Guard against empty versions. Not sure if that's even possible.
+                               Version: crd.Spec.Versions[0].Name,
+                       })
+               }
+       } else {
+               // TODO Detect if the error is because v1 CRDs aren't supported on this cluster.
+               // If so, try the v1beta1
+               crdsv1beta1 := extensionsv1beta1.CustomResourceDefinitionList{}
+
+               err := r.List(ctx, &crdsv1beta1, &client.ListOptions{LabelSelector: crdLabelSelector})
+               if err == nil {
+                       return err
                }
 
-               for _, version := range versions {
+               for _, crd := range crdsv1beta1.Items {
                        tmplGVRs = append(tmplGVRs, schema.GroupVersionResource{
-                               Group:    strings.SplitN(crd.GetName(), ".", 2)[1],
-                               Resource: strings.SplitN(crd.GetName(), ".", 2)[0],
-                               Version:  version.(map[string]interface{})["name"].(string),
+                               Group:    crd.Spec.Group,
+                               Resource: crd.Spec.Names.Plural,
+                               // TODO: Guard against empty versions. Not sure if that's even possible.
+                               Version: crd.Spec.Versions[0].Name,
                        })
                }
        }
diff --git a/main.go b/main.go
index ff27b15..0106603 100644
--- a/main.go
+++ b/main.go
@@ -23,7 +23,10 @@ import (
 
        // to ensure that exec-entrypoint and run can make use of them.
        v1 "k8s.io/api/core/v1"
+       extensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+       extensionsv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
        "k8s.io/apimachinery/pkg/fields"
+       "k8s.io/apimachinery/pkg/labels"
        k8sruntime "k8s.io/apimachinery/pkg/runtime"
        utilruntime "k8s.io/apimachinery/pkg/util/runtime"
        "k8s.io/client-go/kubernetes"
@@ -330,10 +333,23 @@ func getManager(
        )
 
        hubRecorder := eventBroadcaster.NewRecorder(eventsScheme, v1.EventSource{Component: statussync.ControllerName})
+       crdLabelSelector := labels.SelectorFromSet(map[string]string{templatesync.PolicyTypeLabel: "template"})
 
        options.LeaderElectionID = "governance-policy-framework-addon.open-cluster-management.io"
        options.HealthProbeBindAddress = healthAddr
        options.Namespace = tool.Options.ClusterNamespace
+       options.NewCache = cache.BuilderWithOptions(
+               cache.Options{
+                       SelectorsByObject: cache.SelectorsByObject{
+                               &extensionsv1.CustomResourceDefinition{}: {
+                                       Label: crdLabelSelector,
+                               },
+                               &extensionsv1beta1.CustomResourceDefinition{}: {
+                                       Label: crdLabelSelector,
+                               },
+                       },
+               },
+       )
 
        mgr, err := ctrl.NewManager(managedCfg, options)
        if err != nil {
```